pymeasure/instruments/keithley/keithley2182.py

```diff
@@ -0,0 +1,390 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2024 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from pymeasure.instruments import Instrument, Channel
+from pymeasure.instruments.validators import (truncated_range,
+                                              strict_discrete_set,
+                                              strict_range)
+
+from .buffer import KeithleyBuffer
+
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class Keithley2182Channel(Channel):
+    """Implementation of a Keithley 2182 channel.
+
+    Channel 1 is the fundamental measurement channel, while channel 2 provides
+    sense measurements. Channel 2 inputs are referenced to Channel 1 LO.
+    Possible configurations are
+
+    Voltage     (Channel 1)
+    Temperature (Channel 1)
+    Voltage     (Channel 1) and Voltage     (Channel 2)
+    Voltage     (Channel 1) and Temperature (Channel 2)
+    """
+
+    def __init__(self, instrument, id_):
+        """Set max voltage depending on channel."""
+        if id_ == 1:
+            self.voltage_range_values = [0, 120]
+            self.voltage_offset_values = [-120, 120]
+        else:
+            self.voltage_range_values = [0, 12]
+            self.voltage_offset_values = [-12, 12]
+        super().__init__(instrument, id_)
+
+    voltage_range = Channel.control(
+        ":SENS:VOLT:CHAN{ch}:RANG?", ":SENS:VOLT:CHAN{ch}:RANG %g",
+        """Control the positive full-scale measurement voltage range in Volts.
+        Valid ranges are from 0 to 120 V for Ch. 1, and 0 to 12 V for Ch. 2.
+        Auto-range is automatically disabled when this property is set.""",
+        validator=truncated_range,
```

Comment:
I don't know, whether I mentioned it:
Use the `truncated` validators with care!

Imagine, someone writes `voltage_range=200`, which will be effectively only `voltage_range=120`, but the user does not know, that the voltage range is different from his value of 200.

A strict validator will raise an exception, which notifies the user of the wrong value.